internal/jobexecutor/job_executor.go

```diff
@@ -183,16 +184,31 @@ func (e *JobExecutor) execute(ctx context.Context) (res *jobExecutorResult) {
 		return &jobExecutorResult{Err: &rivertype.UnknownJobKindError{Kind: e.JobRow.Kind}, MetadataUpdates: metadataUpdates}
 	}
 
+	// Unmarshal jobs early so that they can be used in places like allowing a
+	// custom args-driven job timeout.
 	if err := e.WorkUnit.UnmarshalJob(); err != nil {
 		return &jobExecutorResult{Err: err, MetadataUpdates: metadataUpdates}
 	}
 
-	doInner := execution.MiddlewareChain(e.GlobalMiddleware, e.WorkUnit.Middleware(), e.WorkUnit.Work, e.JobRow)
+	encodedArgs := e.JobRow.EncodedArgs
+	doInner := execution.Func(func(ctx context.Context) error {
+		// The middleware stack is allowed an opportunity to modify EncodedArgs.
+		// If we do detect a modification, unmarshal args again.
+		if bytes.Compare(encodedArgs, e.JobRow.EncodedArgs) != 0 {
+			if err := e.WorkUnit.UnmarshalJob(); err != nil {
+				return err
+			}
+		}
+
+		return e.WorkUnit.Work(ctx)
+	})
+
+	executeFunc := execution.MiddlewareChain(e.GlobalMiddleware, e.WorkUnit.Middleware(), doInner, e.JobRow)
 	jobTimeout := valutil.FirstNonZero(e.WorkUnit.Timeout(), e.ClientJobTimeout)
 	ctx, cancel := execution.MaybeApplyTimeout(ctx, jobTimeout)
 	defer cancel()
 
-	return &jobExecutorResult{Err: doInner(ctx), MetadataUpdates: metadataUpdates}
+	return &jobExecutorResult{Err: executeFunc(ctx), MetadataUpdates: metadataUpdates}
```

Comment:
Huh, I think I fixed #752 as part of 0132e537ab49246431f4b021f4287ab53ee75f48 / #753 without explicitly calling it out. Looks like it's now applying the timeout to the entire chain.

Reply:
True. Latest push though does change this back to extracting a timeout from within `doInner`, heh.